File: ts/test/entity/country_info/CountryInfoEntity.test.ts


import Path from 'node:path'
import * as Fs from 'node:fs'

import { test, describe, afterEach } from 'node:test'
import assert from 'node:assert'
import { createLiveTransport } from '../../live-runner'
import { runLiveEntity } from '../../live-entity'


import { PublicHolidaySDK, BaseFeature, stdutil } from '../../..'

import {
  envOverride,
  liveClientOptions,
  liveDelay,
  loadEnvLocal,
  makeCtrl,
  makeMatch,
  makeReqdata,
  makeStepData,
  makeValid,
  maybeSkipControl,
} from '../../utility'


// AFTER the imports on purpose: TypeScript hoists `import` above any
// statement in the emitted CommonJS, so a loader placed above them would
// run only after every imported module had already been evaluated - and
// anything reading process.env at module scope would miss these values.
loadEnvLocal(__dirname + '/../../../.env.local')


describe('CountryInfoEntity', async () => {

  // Per-test live pacing. Delay is read from sdk-test-control.json's
  // `test.live.delayMs`; only sleeps when PUBLIC_HOLIDAY_TEST_LIVE=TRUE.
  afterEach(liveDelay('PUBLIC_HOLIDAY_TEST_LIVE'))

  test('instance', async () => {
    const testsdk = PublicHolidaySDK.test()
    const ent = testsdk.CountryInfo()
    assert(null != ent)
  })


  test('basic', async (t) => {

    const live = 'TRUE' === process.env.PUBLIC_HOLIDAY_TEST_LIVE
    for (const op of ['load']) {
      if (!live && maybeSkipControl(t, 'entityOp', 'country_info.' + op, live)) return
    }

    
    const setup = basicSetup()
    if (setup.live) {
      return runLiveEntity(setup, {"active":true,"alias":{"field":{}},"fields":[{"active":true,"name":"borders","req":false,"short":"Bordering countries","type":"`$ARRAY`","index$":0},{"active":true,"name":"commonName","req":false,"short":"Common name of the country","type":"`$STRING`","index$":1},{"active":true,"name":"countryCode","req":false,"short":"ISO 3166-1 alpha-2 country code","type":"`$STRING`","index$":2},{"active":true,"name":"id","req":false,"type":"`$STRING`","index$":3},{"active":true,"name":"officialName","req":false,"short":"Official name of the country","type":"`$STRING`","index$":4},{"active":true,"name":"region","req":false,"short":"Region of the country","type":"`$STRING`","index$":5}],"id":{"field":"id","name":"id"},"name":"country_info","op":{"load":{"input":"data","name":"load","points":[{"active":true,"args":{"params":[{"active":true,"example":"AT","kind":"param","name":"id","orig":"country_code","reqd":true,"type":"`$STRING`","index$":0}]},"contract":{"id":"GET /CountryInfo/{CountryCode}","json":"{\"operationId\":\"getCountryInfo\",\"parameters\":[{\"description\":\"ISO 3166-1 alpha-2 country code\",\"in\":\"path\",\"name\":\"CountryCode\",\"required\":true,\"schema\":{\"example\":\"AT\",\"maxLength\":2,\"minLength\":2,\"type\":\"string\"}}],\"protocol\":\"http\",\"responses\":{\"200\":{\"content\":{\"application/json\":{\"schema\":{\"properties\":{\"borders\":{\"description\":\"Bordering countries\",\"items\":{\"properties\":{\"countryCode\":{\"description\":\"ISO 3166-1 alpha-2 country code\",\"example\":\"AT\",\"type\":\"string\"},\"name\":{\"description\":\"Country name\",\"example\":\"Austria\",\"type\":\"string\"}},\"type\":\"object\"},\"type\":\"array\"},\"commonName\":{\"description\":\"Common name of the country\",\"example\":\"Austria\",\"type\":\"string\"},\"countryCode\":{\"description\":\"ISO 3166-1 alpha-2 country code\",\"example\":\"AT\",\"type\":\"string\"},\"officialName\":{\"description\":\"Official name of the country\",\"example\":\"Republic of Austria\",\"type\":\"string\"},\"region\":{\"description\":\"Region of the country\",\"example\":\"Europe\",\"type\":\"string\"}},\"type\":\"object\"}}},\"description\":\"Successful response with country information\"},\"404\":{\"description\":\"Country not found\"}},\"securitySource\":\"unspecified\"}","source":"openapi3","version":1},"kind":"http","method":"GET","orig":"/CountryInfo/{CountryCode}","rename":{"param":{"CountryCode":"id"}},"segments":[{"lit":"CountryInfo"},{"var":"id"}],"select":{"exist":["id"]},"transform":{"req":"`reqdata`","res":"`body`"},"index$":0}],"key$":"load"}},"relations":{"ancestors":[]},"key$":"country_info","name__orig":"country_info","Name":"CountryInfo","name_":"country_info","name-":"country-info","NAME":"COUNTRY_INFO","index$":1}, {"active":true,"entity":"country_info","key$":"BasicCountryInfoFlow","kind":"basic","name":"BasicCountryInfoFlow","param":{},"step":[{"active":true,"data":{},"input":{"ref":"country_info_ref01","srcdatavar":"country_info_ref01_data","suffix":"_dt0"},"match":{"id":"country_info01"},"op":"load","spec":[],"valid":[{"apply":"TextFieldMark","def":{"mark":"Mark01-country_info_ref01"}}],"index$":0}]}, 'CountryInfo')
    }
    const client = setup.client
    const struct = setup.struct

    const isempty = struct.isempty
    const select = struct.select

    let country_info_ref01_data = Object.values(setup.data.existing.country_info)[0] as any

    // LOAD
    const country_info_ref01_ent = client.CountryInfo()
    const country_info_ref01_match_dt0: any = {}
    country_info_ref01_match_dt0.id = country_info_ref01_data.id
    const country_info_ref01_data_dt0 = (await country_info_ref01_ent.load(country_info_ref01_match_dt0)).data()
    assert(country_info_ref01_data_dt0.id === country_info_ref01_data.id)


  })
})



function basicSetup(extra?: any) {
  // TODO: fix test def options
  const options: any = {} // null

  // TODO: needs test utility to resolve path
  const entityDataFile =
    Path.resolve(__dirname, 
      '../../../../.sdk/test/entity/country_info/CountryInfoTestData.json')

  // TODO: file ready util needed?
  const entityDataSource = Fs.readFileSync(entityDataFile).toString('utf8')

  // TODO: need a xlang JSON parse utility in voxgig/struct with better error msgs
  const entityData = JSON.parse(entityDataSource)

  options.entity = entityData.existing

  let client = PublicHolidaySDK.test(options, extra)
  const struct = client.utility().struct
  const merge = struct.merge
  const transform = struct.transform

  let idmap = transform(
    ['country_info01','country_info02','country_info03'],
    {
      '`$PACK`': ['', {
        '`$KEY`': '`$COPY`',
        '`$VAL`': ['`$FORMAT`', 'upper', '`$COPY`']
      }]
    })

  const env = envOverride({
    'PUBLIC_HOLIDAY_TEST_COUNTRY_INFO_ENTID': idmap,
    'PUBLIC_HOLIDAY_TEST_LIVE': 'FALSE',
    'PUBLIC_HOLIDAY_TEST_EXPLAIN': 'FALSE',
  })

  idmap = env['PUBLIC_HOLIDAY_TEST_COUNTRY_INFO_ENTID']

  const live = 'TRUE' === env.PUBLIC_HOLIDAY_TEST_LIVE

  const transport = createLiveTransport()
  if (live) {
    const rawIds = process.env['PUBLIC_HOLIDAY_TEST_COUNTRY_INFO_ENTID']
    idmap = rawIds && rawIds.trim() ? JSON.parse(rawIds) : {}
    if (!idmap || Array.isArray(idmap) || typeof idmap !== 'object') {
      throw new Error('Live ENTID must be a JSON object')
    }
    client = new PublicHolidaySDK(merge([
      // FIRST, so the generated fields below win: sdk-test-control.json's
      // test.client.options adds to the live client, it does not redirect it.
      liveClientOptions(),
      {
      },
      // 'extra || {}', not a bare 'extra': struct.merge returns UNDEFINED when the
      // last entry is undefined, and basicSetup is normally called with no
      // argument at all - so a bare 'extra' silently discarded the apikey
      // and server values above and handed the SDK undefined. Harmless
      // while there was nothing in that object; not harmless now.
      extra || {},
      { system: { fetch: transport.fetch } }
    ]))
  }

  const setup = {
    idmap,
    env,
    options,
    client,
    struct,
    data: entityData,
    explain: 'TRUE' === env.PUBLIC_HOLIDAY_TEST_EXPLAIN,
    live,
    transport,
    now: Date.now(),
  }

  return setup
}
  

File: go/test/country_info_entity_test.go
package sdktest

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"testing"
	"time"

	sdk "github.com/voxgig-sdk/public-holiday-sdk/go"
	"github.com/voxgig-sdk/public-holiday-sdk/go/core"

	vs "github.com/voxgig-sdk/public-holiday-sdk/go/utility/struct"
)

func TestCountryInfoEntity(t *testing.T) {
	t.Run("instance", func(t *testing.T) {
		testsdk := sdk.TestSDK(nil, nil)
		ent := testsdk.CountryInfo(nil)
		if ent == nil {
			t.Fatal("expected non-nil CountryInfoEntity")
		}
	})

	t.Run("basic", func(t *testing.T) {
		setup := country_infoBasicSetup(nil)
		// Per-op sdk-test-control.json skip — basic test exercises a flow
		// with multiple ops; skipping any op skips the whole flow.
		_mode := "unit"
		if setup.live {
			_mode = "live"
		}
		for _, _op := range []string{"load"} {
			if _shouldSkip, _reason := isControlSkipped("entityOp", "country_info." + _op, _mode); _shouldSkip {
				if _reason == "" {
					_reason = "skipped via sdk-test-control.json"
				}
				t.Skip(_reason)
				return
			}
		}
		// The basic flow consumes synthetic IDs from the fixture. In live mode
		// without an *_ENTID env override, those IDs hit the live API and 4xx.
		if setup.syntheticOnly {
			t.Skip("live entity test uses synthetic IDs from fixture — set PUBLICHOLIDAY_TEST_COUNTRY_INFO_ENTID JSON to run live")
			return
		}
		client := setup.client

		// Bootstrap entity data from existing test data (no create step in flow).
		countryInfoRef01DataRaw := vs.Items(core.ToMapAny(vs.GetPath("existing.country_info", setup.data)))
		var countryInfoRef01Data map[string]any
		if len(countryInfoRef01DataRaw) > 0 {
			countryInfoRef01Data = core.ToMapAny(countryInfoRef01DataRaw[0][1])
		}
		// Discard guards against Go's unused-var check when the flow's steps
		// happen not to consume the bootstrap data (e.g. list-only flows).
		_ = countryInfoRef01Data

		// LOAD
		countryInfoRef01Ent := client.CountryInfo(nil)
		countryInfoRef01MatchDt0 := map[string]any{}
		countryInfoRef01DataDt0Loaded, err := countryInfoRef01Ent.Load(countryInfoRef01MatchDt0, nil)
		if err != nil {
			t.Fatalf("load failed: %v", err)
		}
		if countryInfoRef01DataDt0Loaded == nil {
			t.Fatal("expected load result to be non-nil")
		}

	})
}

func country_infoBasicSetup(extra map[string]any) *entityTestSetup {
	loadEnvLocal()

	_, filename, _, _ := runtime.Caller(0)
	dir := filepath.Dir(filename)

	entityDataFile := filepath.Join(dir, "..", "..", ".sdk", "test", "entity", "country_info", "CountryInfoTestData.json")

	entityDataSource, err := os.ReadFile(entityDataFile)
	if err != nil {
		panic("failed to read country_info test data: " + err.Error())
	}

	var entityData map[string]any
	if err := json.Unmarshal(entityDataSource, &entityData); err != nil {
		panic("failed to parse country_info test data: " + err.Error())
	}

	options := map[string]any{}
	options["entity"] = entityData["existing"]

	client := sdk.TestSDK(options, extra)

	// Generate idmap via transform, matching TS pattern.
	idmap := vs.Transform(
		[]any{"country_info01", "country_info02", "country_info03"},
		map[string]any{
			"`$PACK`": []any{"", map[string]any{
				"`$KEY`": "`$COPY`",
				"`$VAL`": []any{"`$FORMAT`", "upper", "`$COPY`"},
			}},
		},
	)

	// Detect ENTID env override before envOverride consumes it. When live
	// mode is on without a real override, the basic test runs against synthetic
	// IDs from the fixture and 4xx's. Surface this so the test can skip.
	entidEnvRaw := os.Getenv("PUBLICHOLIDAY_TEST_COUNTRY_INFO_ENTID")
	idmapOverridden := entidEnvRaw != "" && strings.HasPrefix(strings.TrimSpace(entidEnvRaw), "{")

	env := envOverride(map[string]any{
		"PUBLICHOLIDAY_TEST_COUNTRY_INFO_ENTID": idmap,
		"PUBLICHOLIDAY_TEST_LIVE":      "FALSE",
		"PUBLICHOLIDAY_TEST_EXPLAIN":   "FALSE",
		"PUBLICHOLIDAY_APIKEY":         "NONE",
	})

	idmapResolved := core.ToMapAny(env["PUBLICHOLIDAY_TEST_COUNTRY_INFO_ENTID"])
	if idmapResolved == nil {
		idmapResolved = core.ToMapAny(idmap)
	}

	if env["PUBLICHOLIDAY_TEST_LIVE"] == "TRUE" {
		mergedOpts := vs.Merge([]any{
			map[string]any{
				"apikey": env["PUBLICHOLIDAY_APIKEY"],
			},
			extra,
		})
		client = sdk.NewPublicHolidaySDK(core.ToMapAny(mergedOpts))
	}

	live := env["PUBLICHOLIDAY_TEST_LIVE"] == "TRUE"
	return &entityTestSetup{
		client:        client,
		data:          entityData,
		idmap:         idmapResolved,
		env:           env,
		explain:       env["PUBLICHOLIDAY_TEST_EXPLAIN"] == "TRUE",
		live:          live,
		syntheticOnly: live && !idmapOverridden,
		now:           time.Now().UnixMilli(),
	}
}
